# 两种形式

### callback

回调形式: 组件采取方法回调的方式通知调用者,将判断的权利交给用户，更加灵活
> 采用`build`方法进行构造，用户实现`Retryable`接口中的方法

`new RetryBuilder().build().proceed((round, nanos) -> "success");`

### condition

条件形式: 组件不会回调任何方法(包括`whenError`,`whenFinish`,`postCondition`,`preCondition`)`whenTimeout`除外，只根据`failOn`以及`continueOn`进行循环控制
> 采用`buildCondition`方法进行构造，用户实现`ConditionRetryable`接口中的`get`方法

``` java
Class[] failOn = {IllegalAccessException.class};
Class[] continueOn = {IllegalArgumentException.class};
new RetryBuilder()
        .failOn(failOn)
        .continueOn(continueOn)
        .buildCondition()
        .proceed(() -> "success");
```
- 当匹配到`failOn`异常时，抛出ProceedException,cause为具体异常
- 当匹配到`continueOn`异常时，继续循环。未匹配到时抛出ProceedException,cause为具体异常
- 先判断`failOn`后判断`continueOn`
- 如果`failOn`,`continueOn`都未匹配上时继续循环

``` java
例1
Class[] failOn = {IllegalAccessException.class};
当发生IllegalAccessException时，程序将抛出ProceedException（满足failOn）
当发生 IllegalAccessException外的其他异常时，将继续运行
```

``` java
例2
Class[] continueOn = {IllegalAccessException.class,RuntimeException.class};
当发生IllegalAccessException时，将继续运行（满足continueOn）
当发生 IllegalAccessException，RuntimeException 外的其他异常时，程序将抛出ProceedException
```

``` java
例3
Class[] failOn = {RuntimeException.class};
Class[] continueOn = {RuntimeException.class};
当发生RuntimeException时，程序将抛出ProceedException(先判断`failOn`后判断`continueOn`)
```